clouddriver-aws/src/main/groovy/com/netflix/spinnaker/clouddriver/aws/provider/agent/AmazonApplicationLoadBalancerCachingAgent.groovy

```diff
@@ -95,25 +132,103 @@ class AmazonApplicationLoadBalancerCachingAgent extends AbstractAmazonLoadBalanc
-      def loadBalancing = amazonClientProvider.getAmazonElasticLoadBalancingV2(account, region, true)
+    def loadBalancing = amazonClientProvider.getAmazonElasticLoadBalancingV2(account, region, true)
+
+    LoadBalancer loadBalancer = metricsSupport.readData {
       try {
         return loadBalancing.describeLoadBalancers(
           new DescribeLoadBalancersRequest().withNames([data.loadBalancerName as String])
-        ).loadBalancers
+        ).loadBalancers.get(0)
       } catch (LoadBalancerNotFoundException ignored) {
+        return null
+      }
+    }
+    if (loadBalancer == null) { loadBalancer = [:] }
+
+    List<TargetGroup> targetGroups = metricsSupport.readData {
+      try {
+        return loadBalancing.describeTargetGroups(
+          new DescribeTargetGroupsRequest().withLoadBalancerArn(loadBalancer?.loadBalancerArn)
+        ).targetGroups
+      } catch (LoadBalancerNotFoundException ignored) {
+        return []
+      } catch (TargetGroupNotFoundException ignored) {
         return []
       }
     }
 
-    def cacheResult = metricsSupport.transformData { buildCacheResult(loadBalancers, [:], System.currentTimeMillis(), []) }
+    // Get all the target group health and attributes
+    Map<String, List<TargetHealthDescription>> targetGroupArnToHealths = metricsSupport.readData {
+      targetGroups.collectEntries { targetGroup ->
+        List<TargetHealthDescription> targetHealthDescriptions = loadBalancing.describeTargetHealth(
+          new DescribeTargetHealthRequest().withTargetGroupArn(targetGroup.targetGroupArn)
+        ).targetHealthDescriptions
+        return [(targetGroup.targetGroupArn): targetHealthDescriptions]
+      }
+    }
+
+    Map<String, List<TargetGroupAttribute>> targetGroupArnToAttributes = metricsSupport.readData {
+      targetGroups.collectEntries { targetGroup ->
+        List<TargetGroupAttribute> targetGroupAttributes = loadBalancing.describeTargetGroupAttributes(
+          new DescribeTargetGroupAttributesRequest().withTargetGroupArn(targetGroup.targetGroupArn)
+        ).attributes
+        return [(targetGroup.targetGroupArn): targetGroupAttributes]
+      }
+    }
+
+    Map<String, List<Listener>> loadBalancerArnToListeners = [:]
+    Map<Listener, List<Rule>> listenerToRules = new HashMap<>()
+
+    // Add the listeners
+    List<Listener> listenerData = new ArrayList<>()
+    try {
+      DescribeListenersRequest describeListenersRequest =
+        new DescribeListenersRequest(loadBalancerArn: loadBalancer.loadBalancerArn)
+      while (true) {
+        DescribeListenersResult result = loadBalancing.describeListeners(describeListenersRequest)
+        listenerData.addAll(result.listeners)
+        if (result.nextMarker) {
+          describeListenersRequest.withMarker(result.nextMarker)
+        } else {
+          break
+        }
+      }
+    } catch (LoadBalancerNotFoundException ignore) {
+      // this is acceptable since we may be waiting for the caches to catch up
```

Comment:
confused ... what are we waiting to catch up? These calls should be going directly against AWS.